src/common/acl_principal.c

```diff
@@ -0,0 +1,344 @@
+	char		*buf = NULL;
+	char		*new_buf = NULL;
+	size_t		buflen = DEFAULT_BUF_LEN;
+
+	if (name == NULL) {
+		D_INFO("name pointer was NULL!\n");
+		return -DER_INVAL;
+	}
+
+	/*
+	 * No platform-agnostic way to fetch the max buflen - so let's try a
+	 * sane value and double until it's big enough.
+	 */
+	do {
+		D_REALLOC(new_buf, buf, buflen);
+		if (new_buf == NULL) {
+			D_ERROR("Couldn't allocate memory for getgrgid_r\n");
+			return -DER_NOMEM;
+		}
+
+		buf = new_buf;
+
+		rc = getgrgid_r(gid, &grp, buf, buflen, &result);
+
+		buflen *= 2;
+	} while (rc == ERANGE);
+
+	if (rc != 0) {
+		D_ERROR("Error from getgrgid_r: %d\n", rc);
+		D_GOTO(out, rc = d_errno2der(rc));
+	}
+
+	if (result == NULL) {
+		D_INFO("No group for gid %u\n", gid);
+		D_GOTO(out, rc = -DER_NONEXIST);
+	}
+
+	rc = local_name_to_principal_name(result->gr_name, name);
+
+out:
+	D_FREE(buf);
+	return rc;
+}
+
+/*
+ * Extracts the user/group name from the "name@[domain]" formatted principal
+ * string.
+ */
+static int
+get_id_name_from_principal(const char *principal, char *name)
+{
+	int num_matches;
+
+	if (!daos_acl_principal_is_valid(principal)) {
+		D_INFO("Invalid name format\n");
+		return -DER_INVAL;
+	}
+
+	num_matches = sscanf(principal, "%[^@]", name);
+	if (num_matches == 0) {
+		/*
+		 * This is a surprise - if it's formatted properly, we should
+		 * be able to extract the name.
+		 */
+		D_ERROR("Couldn't extract ID name from '%s'\n", principal);
+		return -DER_INVAL;
+	}
+
+	return 0;
+}
+
+int
+daos_acl_principal_to_uid(const char *principal, uid_t *uid)
+{
+	char		username[DAOS_ACL_MAX_PRINCIPAL_BUF_LEN];
+	char		*buf = NULL;
+	char		*new_buf = NULL;
+	size_t		buflen = DEFAULT_BUF_LEN;
+	struct passwd	passwd;
+	struct passwd	*result = NULL;
+	int		rc;
+
+	if (uid == NULL) {
+		D_INFO("NULL uid pointer\n");
+		return -DER_INVAL;
+	}
+
+	rc = get_id_name_from_principal(principal, username);
+	if (rc != 0)
+		return rc;
+
+	do {
+		D_REALLOC(new_buf, buf, buflen);
+		if (new_buf == NULL) {
+			D_ERROR("Couldn't alloc buffer for getpwnam_r\n");
+			return -DER_NOMEM;
+		}
+
+		buf = new_buf;
+
+		rc = getpwnam_r(username, &passwd, buf, buflen, &result);
+
+		buflen *= 2;
+	} while (rc == ERANGE);
+
+	if (rc != 0) {
+		D_ERROR("Error from getpwnam_r: %d\n", rc);
+		D_GOTO(out, rc = d_errno2der(rc));
+	}
+
+	if (result == NULL) {
+		D_INFO("User '%s' not found\n", username);
+		D_GOTO(out, rc = -DER_NONEXIST);
+	}
+
+	*uid = result->pw_uid;
+	rc = 0;
+out:
+	D_FREE(buf);
+	return rc;
+}
+
+int
+daos_acl_principal_to_gid(const char *principal, gid_t *gid)
+{
+	char		grpname[DAOS_ACL_MAX_PRINCIPAL_BUF_LEN];
+	char		*buf = NULL;
+	char		*new_buf = NULL;
+	size_t		buflen = DEFAULT_BUF_LEN;
+	struct group	grp;
+	struct group	*result = NULL;
+	int		rc;
+
+	if (gid == NULL) {
+		D_INFO("NULL gid pointer\n");
+		return -DER_INVAL;
+	}
+
+	rc = get_id_name_from_principal(principal, grpname);
+	if (rc != 0)
+		return rc;
+
+	do {
+		D_REALLOC(new_buf, buf, buflen);
```

Comment:
And here.

You might consider commonizing this loop between the different cases since it's repeating the same generic code.

Reply:
Created a macro that could be reused in all these cases. Not totally sure I like the level of indirection that adds, but I think it's a simple enough loop in this case.